### PR TITLE
Update pow chain deposit contract and tests

### PR DESCRIPTION
**Pull request by NIC619**
- [x] Update deposit contract with the one from the [spec](https://github.com/ethereum/eth2.0-specs/blob/9e74043a1737b1a792e259ef93c995e0634df086/specs/core/0_beacon-chain.md#ethereum-10-deposit-contract)
- [x] Update Vyper version
- [x] Update deposit tests
    - deposit amount
    - deposit log
    - receipt tree
- [x] Add `ChainStart` test

### CODE REVIEW

File: tests/contracts/test_deposit.py [hunk 1]
```diff
@@ -1,22 +1,32 @@
...
 @pytest.mark.parametrize(
     'success,amount_deposit',
     [
         (True, 32),
-        (False, 31),
-        (False, 33),
-        (False, 0)
+        (True, 1),
+        (False, 0),
+        (False, 33)
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]: What do you think about defining the constants in `conftest.py`?
```
MIN_DEPOSIT: constant(uint256) = 1  # ETH
MAX_DEPOSIT: constant(uint256) = 32  # ETH
GWEI_PER_ETH: constant(uint256) = 1000000000  # 10**9
CHAIN_START_FULL_DEPOSIT_THRESHOLD: constant(uint256) = 16384  # 2**14
DEPOSIT_CONTRACT_TREE_DEPTH: constant(uint256) = 32
```

Comment (hwwhww) [inline tests/contracts/test_deposit.py]: Then this line would be:
```suggestion
        (False, MAX_DEPOSIT + 1)
```

File: tests/contracts/test_deposit.py [hunk 2]
```diff
@@ -1,22 +1,32 @@
...
+
+
 @pytest.mark.parametrize(
     'success,amount_deposit',
     [
         (True, 32),
-        (False, 31),
-        (False, 33),
-        (False, 0)
+        (True, 1),
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]:
```suggestion
        (True, MIN_DEPOSIT),
```

File: tests/contracts/test_deposit.py [hunk 3]
```diff
@@ -25,37 +35,67 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
...
-    log_filter = registration_contract.events.Deposit.createFilter(
+    log_filter = modified_registration_contract.events.ChainStart.createFilter(
         fromBlock='latest')
 
-    registration_contract.functions.deposit(
-        b'\x00'*32,
-        43,
-        a0,
-        b'\x00'*32).transact({"value": w3.toWei(32, "ether")})
+    max_deposit_amount = 32 * 10**9
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]:
```suggestion
    max_deposit_amount = MAX_DEPOSIT * eth_utils.denoms.gwei  # in gwei
```

File: tests/contracts/test_deposit.py [hunk 4]
```diff
@@ -25,37 +35,67 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
...
+    # Next make 4 deposit with value MAX_DEPOSIT
+    for i in range(2, 6):
+        deposit_parameters = i.to_bytes(1, 'big') * 100
+        modified_registration_contract.functions.deposit(
+            deposit_parameters).transact({"value": w3.toWei(max_deposit_amount, "gwei")})
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]: nitpick
```suggestion
            deposit_parameters,
        ).transact({"value": w3.toWei(max_deposit_amount, "gwei")})
```

File: tests/contracts/test_deposit.py [hunk 5]
```diff
@@ -25,37 +35,67 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
...
+        logs = log_filter.get_new_entries()
+        # ChainStart event should not be triggered
+        assert len(logs) == 0
 
+    # Make 1 more deposit with value MAX_DEPOSIT to trigger ChainStart event
+    deposit_parameters = b'\x06' * 100
+    modified_registration_contract.functions.deposit(
+        deposit_parameters).transact({"value": w3.toWei(max_deposit_amount, "gwei")})
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]: nitpick
```suggestion
        deposit_parameters,
    ).transact({"value": w3.toWei(max_deposit_amount, "gwei")})
```

**Comment by NIC619** [comment]
@hwwhww 
> 1. Could you meter the gas cost of calling deposit()?

First `deposit` transaction takes around 980k gas while the subsequent ones take around 480k gas. 

> 2. More test cases:
Try when CHAIN_START_FULL_DEPOSIT_THRESHOLD=16384: we might need to see the edge case? (<-- not quite sure)

`modified_registration_contract` fixture has `CHAIN_START_FULL_DEPOSIT_THRESHOLD` set to 5 and in `test_chain_start` test case it verifies that only with enough `full_deposit_count` will it trigger `ChainStart` event.
My assumption was that from evm's point of view it does not matter if `CHAIN_START_FULL_DEPOSIT_THRESHOLD` is set to 16384 or 5(unless we are testing if the contract will break if it stores enormous amount of validator data?). But it may not be the case from the contract logic's point of view. Please let me know if that does not make sense or there are other edge cases that comes up your mind.
Another problem with setting `CHAIN_START_FULL_DEPOSIT_THRESHOLD` too high is that currently it takes about 6 seconds to complete one `deposit` transaction in test.

File: tests/contracts/test_deposit.py [hunk 6]
```diff
@@ -25,37 +43,76 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
...
-        a0,
-        b'\x00'*32).transact({"value": w3.toWei(32, "ether")})
+    leaf_nodes = []
+    for i in range(1, 10):
+        deposit_input = i.to_bytes(1, 'big') * 100
+        tx_hash = registration_contract.functions.deposit(
+            deposit_input,
+        ).transact({"value": w3.toWei(deposit_amount, "gwei")})
+        receipt = w3.eth.getTransactionReceipt(tx_hash)
+        print("deposit transaction consumes %d gas", receipt['gasUsed'])
```
Comment (hwwhww) [inline tests/contracts/test_deposit.py]:
```suggestion
        print("deposit transaction consumes %d gas" % receipt['gasUsed'])
```

File: tests/contracts/test_deposit.py [hunk 7]
```diff
@@ -25,37 +43,76 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
         )
 
 
-def test_no_reuse_of_pubkey(registration_contract, a0, w3, assert_tx_failed):
+def test_deposit_log(registration_contract, a0, w3):
```
Comment (djrtwo) [inline tests/contracts/test_deposit.py]: * should parametrize with 2 or 3 different deposit amounts.
* should run a test that issues two deposits to ensure that `previous_receipt_root` field becomes non-zero and `deposit_count` increments

File: tests/contracts/test_deposit.py [hunk 8]
```diff
@@ -25,37 +43,76 @@ def test_deposit(registration_contract, a0, w3, success, amount_deposit, assert_
...
+    assert log['previous_receipt_root'] == b'\x00' * 32
+    assert log['data'] == amount_bytes8 + timestamp_bytes8 + deposit_input
+    assert log['deposit_count'] == 0
 
 
-def test_log_is_captured(registration_contract, a0, w3):
-    log_filter = registration_contract.events.Deposit.createFilter(
-        fromBlock='latest')
+def test_reciept_tree(registration_contract, w3, assert_tx_failed):
+    deposit_amount = MAX_DEPOSIT * eth_utils.denoms.gwei
```
Comment (djrtwo) [inline tests/contracts/test_deposit.py]: consider making `deposit_amount` dynamic based upon `i` in the loop

```python
deposit_amount = i * 2 * eth_utils.denoms.gwei
```

**Comment by NIC619** [comment]
@hwwhww @djrtwo 
Add `TWO_TO_POWER_OF_TREE_DEPTH` and use random `CHAIN_START_FULL_DEPOSIT_THRESHOLD` and random deposit amount in tests.

**Comment by djrtwo** [comment]
lgtm!
